diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-base-11.3.0/apps/wms_inoapi.c asterisk-11.3.0/apps/wms_inoapi.c
--- asterisk-base-11.3.0/apps/wms_inoapi.c	2014-03-17 16:20:51.544560561 +0000
+++ asterisk-11.3.0/apps/wms_inoapi.c	2014-03-20 17:18:27.547860000 +0000
@@ -1456,6 +1456,7 @@
    trs_reply.event = SEND_FAX_COMPLETED;
    trs_reply.error = TRUE;
 
+   trs_reply.body.fax_reply.status = -1; // FAX ERROR
    trs_reply.body.fax_reply.event = SEND_FAX_COMPLETED;      // event returned
    trs_reply.body.fax_reply.chID = chan_vars->trs_reply.chan_id;   // channel ID
    trs_reply.body.fax_reply.format = FAX_FORMAT_TIFF;     // fax storage format
@@ -1560,6 +1561,7 @@
    trs_reply.event = SEND_FAX_COMPLETED;
    trs_reply.error = TRUE;
 
+   trs_reply.body.fax_reply.status = -1; // FAX ERROR
    trs_reply.body.fax_reply.event = SEND_FAX_COMPLETED;      // event returned
    trs_reply.body.fax_reply.chID = chan_vars->trs_reply.chan_id;   // channel ID
    trs_reply.body.fax_reply.format = FAX_FORMAT_TIFF;     // fax storage format
@@ -1593,6 +1595,31 @@
       if (option_verbose >= 3)
          ast_verbose(VERBOSE_PREFIX_3 "send_fax_thread: Exiting SendFax on channel %s\n", ast_channel_name(chan));
 
+
+      if(!ast_test_flag(ast_channel_inoflags(chan), FAX_PHASE_B))
+      {
+         // event changed, maybe a release_enpoind or stop_operation
+         // return without sending an event
+         if (option_verbose >= 3)
+            ast_verbose(VERBOSE_PREFIX_3 "send_fax_thread: Fax didn't reach phase B on channel %s\n", ast_channel_name(chan));
+
+         // Clear FAX_TONE_DETECTED flag in case it was activated
+         ast_channel_lock(chan);
+         ast_clear_flag(ast_channel_inoflags(chan), FAX_TONE_CNG_DETECTED);
+         ast_clear_flag(ast_channel_inoflags(chan), FAX_TONE_CED_DETECTED);
+         ast_channel_unlock(chan);
+
+         trs_reply.body.fax_reply.status = 0; // SUCCESS
+         trs_reply.body.error_value = 3337; // Far end is not able to receive
+
+         // send event
+         asteriskSendReply(&trs_reply, chan_vars);
+
+         chan_vars->fax.enabled = FALSE;
+
+         return NULL;
+      }
+
       if(ast_test_flag(ast_channel_inoflags(chan), EXEC_APP_EXIT))
       {
          // event changed, maybe a release_enpoind or stop_operation
@@ -1621,7 +1648,7 @@
             if((s_tmp2 = pbx_builtin_getvar_helper(chan, "FAXERRORVALUE")))
             {
                sscanf(s_tmp2,"%d", &s_error);
-               ast_log(LOG_ERROR, "send_fax_thread: FAXERRORVALUE is s_error: %d + FAX_ERROR_MASK: %d\n", s_error, FAX_ERROR_MASK);
+               ast_log(LOG_ERROR, "send_fax_thread: FAXERRORVALUE is %d\n", s_error + FAX_ERROR_MASK);
                trs_reply.body.fax_reply.status = s_error + FAX_ERROR_MASK; // FAILED
                trs_reply.body.error_value = s_error + FAX_ERROR_MASK; // FAILED
             }
@@ -1665,7 +1692,7 @@
                      if((s_tmp = pbx_builtin_getvar_helper(chan, "FAXERRORVALUE")))
                      {
                         sscanf(s_tmp,"%d", &s_error);
-                        ast_log(LOG_ERROR, "send_fax_thread: FAXERRORVALUE is s_error: %d + FAX_ERROR_MASK: %d\n", s_error, FAX_ERROR_MASK);
+                        ast_log(LOG_ERROR, "send_fax_thread: FAXERRORVALUE is %d\n", s_error + FAX_ERROR_MASK);
                         trs_reply.body.fax_reply.status = s_error + FAX_ERROR_MASK; // FAILED
                         trs_reply.body.error_value = s_error + FAX_ERROR_MASK; // FAILED
                      }
@@ -1762,6 +1789,7 @@
    trs_reply.event = RECEIVE_FAX_COMPLETED;
    trs_reply.error = TRUE;
 
+   trs_reply.body.fax_reply.status = -1; // FAX ERROR
    trs_reply.body.fax_reply.event = RECEIVE_FAX_COMPLETED;      // event returned
    trs_reply.body.fax_reply.chID = chan_vars->trs_reply.chan_id;       // channel ID
    trs_reply.body.fax_reply.format = FAX_FORMAT_TIFF;     // fax storage format
@@ -1860,6 +1888,7 @@
    trs_reply.event = RECEIVE_FAX_COMPLETED;
    trs_reply.error = TRUE;
 
+   trs_reply.body.fax_reply.status = -1; // FAX ERROR
    trs_reply.body.fax_reply.event = RECEIVE_FAX_COMPLETED;      // event returned
    trs_reply.body.fax_reply.chID = chan_vars->trs_reply.chan_id;   // channel ID
    trs_reply.body.fax_reply.format = FAX_FORMAT_TIFF;     // fax storage format
@@ -1892,6 +1921,30 @@
       if (option_verbose >= 3)
          ast_verbose(VERBOSE_PREFIX_3 "receive_fax_thread: Exiting ReceiveFax on channel %s\n", ast_channel_name(chan));
 
+      if(!ast_test_flag(ast_channel_inoflags(chan), FAX_PHASE_B))
+      {
+         // event changed, maybe a release_enpoind or stop_operation
+         // return without sending an event
+         ast_log(LOG_DEBUG, "receive_fax_thread: Fax didn't reach phase B on ch %d on channel %s\n",
+               chan_vars->trs_reply.chan_id, ast_channel_name(chan));
+
+         // Clear FAX_TONE_DETECTED flag in case it was activated
+         ast_channel_lock(chan);
+         ast_clear_flag(ast_channel_inoflags(chan), FAX_TONE_CED_DETECTED);
+         ast_clear_flag(ast_channel_inoflags(chan), FAX_TONE_CNG_DETECTED);
+         ast_channel_unlock(chan);
+
+         trs_reply.body.fax_reply.status = 0; // SUCCESS
+         trs_reply.body.error_value = 3338; // Far end is not able to transmit
+
+         // send event
+         asteriskSendReply(&trs_reply, chan_vars);
+
+         chan_vars->fax.enabled = FALSE;
+
+         return NULL;
+      }
+
       if(ast_test_flag(ast_channel_inoflags(chan), EXEC_APP_EXIT))
       {
          // event changed, maybe a release_enpoind or stop_operation
@@ -1921,7 +1974,7 @@
             if((s_tmp2 = pbx_builtin_getvar_helper(chan, "FAXERRORVALUE")))
             {
                sscanf(s_tmp2,"%d", &s_error);
-               ast_log(LOG_ERROR, "receive_fax_thread: FAXERRORVALUE is s_error: %d + FAX_ERROR_MASK: %d\n", s_error, FAX_ERROR_MASK);
+               ast_log(LOG_ERROR, "receive_fax_thread: FAXERRORVALUE is %d\n", s_error + FAX_ERROR_MASK);
                trs_reply.body.fax_reply.status = s_error + FAX_ERROR_MASK; // FAILED
                trs_reply.body.error_value = s_error + FAX_ERROR_MASK; // FAILED
             }
@@ -1963,7 +2016,7 @@
                      if((s_tmp = pbx_builtin_getvar_helper(chan, "FAXERRORVALUE")))
                      {
                         sscanf(s_tmp,"%d", &s_error);
-                        ast_log(LOG_ERROR, "receive_fax_thread: FAXERRORVALUE is s_error: %d + FAX_ERROR_MASK: %d\n", s_error, FAX_ERROR_MASK);
+                        ast_log(LOG_ERROR, "receive_fax_thread: FAXERRORVALUE is %d\n", s_error + FAX_ERROR_MASK);
                         trs_reply.body.fax_reply.status = s_error + FAX_ERROR_MASK; // FAILED
                         trs_reply.body.error_value = s_error + FAX_ERROR_MASK; // FAILED
                      }
diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-base-11.3.0/include/asterisk/channel.h asterisk-11.3.0/include/asterisk/channel.h
--- asterisk-base-11.3.0/include/asterisk/channel.h	2014-03-17 16:20:45.872561562 +0000
+++ asterisk-11.3.0/include/asterisk/channel.h	2014-03-19 17:06:17.948645000 +0000
@@ -195,6 +195,7 @@
 
 #define FAX_ACTIVE            (1 << 19)
 #define FAX_HANGUP            (1 << 20)
+#define FAX_PHASE_B           (1 << 21)
 
 enum ast_bridge_result {
 	AST_BRIDGE_COMPLETE = 0,
diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-base-11.3.0/res/res_fax.c asterisk-11.3.0/res/res_fax.c
--- asterisk-base-11.3.0/res/res_fax.c	2014-03-17 16:22:21.135810353 +0000
+++ asterisk-11.3.0/res/res_fax.c	2014-03-20 09:33:52.046755000 +0000
@@ -1634,7 +1634,8 @@
 			}
 
 			if (!(frame = ast_read(chan))) {
-				ast_log(LOG_ERROR, "error reading frame while generating CED tone on %s\n", ast_channel_name(chan));
+			   if(!ast_check_hangup(chan))
+			      ast_log(LOG_ERROR, "error reading frame while generating CED tone on %s\n", ast_channel_name(chan));
 				ast_playtones_stop(chan);
 				return -1;
 			}
@@ -1956,7 +1957,12 @@
 			fax_session_release(s, token);
 			ao2_ref(s, -1);
 			ao2_ref(details, -1);
-			ast_log(LOG_ERROR, "error initializing channel '%s' in T.38 mode\n", ast_channel_name(chan));
+
+         if(ast_check_hangup(chan))
+            ast_log(LOG_WARNING, "FAX Release on channel %s\n", ast_channel_name(chan));
+         else
+            ast_log(LOG_ERROR, "error initializing channel '%s' in T.38 mode\n", ast_channel_name(chan));
+
 			return -1;
 		}
 	}
@@ -2054,7 +2060,8 @@
 		}
 
 		if (!(frame = ast_read(chan))) {
-			ast_log(LOG_ERROR, "error reading frame while generating CNG tone on %s\n", ast_channel_name(chan));
+		   if(!ast_check_hangup(chan))
+		      ast_log(LOG_ERROR, "error reading frame while generating CNG tone on %s\n", ast_channel_name(chan));
 			ast_playtones_stop(chan);
 			return -1;
 		}
@@ -2471,7 +2478,13 @@
 			fax_session_release(s, token);
 			ao2_ref(s, -1);
 			ao2_ref(details, -1);
-			ast_log(LOG_ERROR, "error initializing channel '%s' in T.38 mode\n", ast_channel_name(chan));
+
+
+         if(ast_check_hangup(chan))
+            ast_log(LOG_WARNING, "FAX Release on channel %s\n", ast_channel_name(chan));
+         else
+            ast_log(LOG_ERROR, "error initializing channel '%s' in T.38 mode\n", ast_channel_name(chan));
+
 			return -1;
 		}
 	} else {
diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-base-11.3.0/res/res_fax_spandsp.c asterisk-11.3.0/res/res_fax_spandsp.c
--- asterisk-base-11.3.0/res/res_fax_spandsp.c	2014-03-17 16:22:21.143810407 +0000
+++ asterisk-11.3.0/res/res_fax_spandsp.c	2014-03-19 17:07:48.790559000 +0000
@@ -365,6 +365,7 @@
    if (completion_code == T30_ERR_OK || completion_code == 131 || completion_code == 128) {
       ast_string_field_set(s->details, resultstr, "");
       ast_set_flag(ast_channel_inoflags(s->chan), FAX_ACTIVE);
+      ast_set_flag(ast_channel_inoflags(s->chan), FAX_PHASE_B);
    } else {
       ast_string_field_set(s->details, resultstr, t30_completion_code_to_str(completion_code));
       s_result = completion_code;
@@ -565,7 +566,7 @@
       pbx_builtin_setvar_helper(s->chan, "FAXTRSIZE", s_size);
    }
 
-   if(s_chid)
+   if(s_chid && ast_test_flag(ast_channel_inoflags(s->chan), FAX_PHASE_B))
    {
       manager_event(EVENT_FLAG_CALL,
                        "FaxPhaseE",
