diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-base-11.3.0/apps/wms_inoapi.c asterisk-11.3.0/apps/wms_inoapi.c
--- asterisk-base-11.3.0/apps/wms_inoapi.c	2022-07-07 12:02:13.080979051 +0100
+++ asterisk-11.3.0/apps/wms_inoapi.c	2022-07-07 12:03:11.015445648 +0100
@@ -420,6 +420,7 @@
  *****************************************************************************/
 void release_endpoint(struct ast_channel *chan, cmd_ast trs_cmd, ast_vars *chan_vars)
 {
+   char bye_body[256];
    if (option_verbose >= 3)
       ast_verbose(VERBOSE_PREFIX_3 "Release on channel %s\n", ast_channel_name(chan));
 
@@ -438,7 +439,8 @@
       }
       if (strlen(trs_cmd.al_char_val1) > 0 )
       {
-         ast_channel_bye_msg_body_set(chan, trs_cmd.al_char_val1);
+         snprintf(bye_body, sizeof(bye_body), "%s\r\n", trs_cmd.al_char_val1);
+         ast_channel_bye_msg_body_set(chan, bye_body);
       }
       if (trs_cmd.al_int_val2 >= 100)
          ast_channel_hangupcause_set(chan, trs_cmd.al_int_val2);
diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-base-11.3.0/channels/chan_sip.c asterisk-11.3.0/channels/chan_sip.c
--- asterisk-base-11.3.0/channels/chan_sip.c	2022-07-07 12:02:16.586067727 +0100
+++ asterisk-11.3.0/channels/chan_sip.c	2022-07-07 12:03:03.515255525 +0100
@@ -4816,7 +4816,7 @@
 /*! \brief add a blank line if no body or SIP request is a BYE message - RFC 5552 4.2 */
 static void add_blank(struct sip_request *req)
 {
-	if (!req->lines || (req->method == SIP_BYE && req->embedded_isup != TRUE)) {
+	if (!req->lines) {
 		/* Add extra empty return. add_header() reserves 4 bytes so cannot be truncated */
 		ast_str_append(&req->data, 0, "\r\n");
 	}
