diff -Naur asterisk-base-1.8.3.2/apps/wms_inoapi.c asterisk-1.8.3.2/apps/wms_inoapi.c
--- asterisk-base-1.8.3.2/apps/wms_inoapi.c	2012-12-03 09:17:34.379543834 +0000
+++ asterisk-1.8.3.2/apps/wms_inoapi.c	2012-12-03 09:36:29.714321000 +0000
@@ -308,6 +308,9 @@
 	{
 		trs_reply.error = FALSE;
 		trs_reply.body.g_ret_value = pbx_exec(chan, app, trs_cmd.al_char_val2);
+		
+		if (!strcasecmp(trs_cmd.al_char_val1, "transfer"))
+			chan_vars->shutdown = 1;
 	}
 	else // application doesn't exist
 	{
diff -Naur asterisk-base-1.8.3.2/channels/chan_sip.c asterisk-1.8.3.2/channels/chan_sip.c
--- asterisk-base-1.8.3.2/channels/chan_sip.c	2012-12-03 09:17:28.996947239 +0000
+++ asterisk-1.8.3.2/channels/chan_sip.c	2012-12-03 09:35:20.575609000 +0000
@@ -21930,10 +21930,17 @@
 			success = FALSE;
 			break;
 		default:
+		
+			if (respcode == 100) {
+				success = TRUE;
+				break;
+			}
+			
 			if (respcode < 200) {
 				/* ignore provisional responses */
 				success = -1;
 			} else {
+			
 				ast_log(LOG_NOTICE, "Got unknown code '%d' in NOTIFY in response to REFER.\n", respcode);
 				success = FALSE;
 			}
