diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-base-11.3.0/apps/wms_inoapi.c asterisk-11.3.0/apps/wms_inoapi.c
--- asterisk-base-11.3.0/apps/wms_inoapi.c	2021-02-17 16:37:58.287696350 +0000
+++ asterisk-11.3.0/apps/wms_inoapi.c	2021-02-17 17:34:57.297439413 +0000
@@ -222,14 +222,15 @@
 
 			gettimeofday(&chan_vars->asr.asr_txcore, NULL);
 
-			ast_format_set(&rtp_ulaw, AST_FORMAT_ULAW, 0);
+			//ast_format_set(&rtp_ulaw, AST_FORMAT_ULAW, 0);
 			ast_format_set(&rtp_alaw, AST_FORMAT_ALAW, 0);
 
-			if ((!ast_format_cap_iscompatible(ast_channel_nativeformats(chan), &rtp_ulaw)) && ((!ast_format_cap_iscompatible(ast_channel_nativeformats(chan), &rtp_alaw))))
+			//if ((!ast_format_cap_iscompatible(ast_channel_nativeformats(chan), &rtp_ulaw)) && ((!ast_format_cap_iscompatible(ast_channel_nativeformats(chan), &rtp_alaw))))
+			if (!ast_format_cap_iscompatible(ast_channel_nativeformats(chan), &rtp_alaw))
 			{
 			   asr_cap = ast_format_cap_alloc();
 			   ast_format_cap_add(asr_cap, &rtp_alaw);
-			   ast_format_cap_add(asr_cap, &rtp_ulaw);
+			   //ast_format_cap_add(asr_cap, &rtp_ulaw);
 
 				res = ast_translator_best_choice(asr_cap, ast_channel_nativeformats(chan), &rtp_alaw, &src);
 				if (chan_vars->asr.asr_writetrans) ast_translator_free_path(chan_vars->asr.asr_writetrans);
