diff -Naur asterisk-base-1.8.3.2/channels/chan_sip.c asterisk-1.8.3.2/channels/chan_sip.c
--- asterisk-base-1.8.3.2/channels/chan_sip.c   2013-01-08 18:32:44.792938159 +0000
+++ asterisk-1.8.3.2/channels/chan_sip.c        2013-01-11 18:12:24.670676000 +0000
@@ -22650,6 +22650,8 @@
                                ast_queue_control(p->owner, AST_CONTROL_SRCUPDATE);
                        } else {
                                p->jointcapability = p->capability;
+                               /* If other party has no SDP, then it could have the same noncodeccapability */
+                               p->jointnoncodeccapability = p->noncodeccapability;
                                ast_debug(1, "Hm....  No sdp for the moment\n");
                                /* Some devices signal they want to be put off hold by sending a re-invite
                                   *without* an SDP, which is supposed to mean "Go back to your state"
@@ -22732,6 +22734,8 @@
                        }
                } else {        /* No SDP in invite, call control session */
                        p->jointcapability = p->capability;
+                       /* If other party has no SDP, then it could have the same noncodeccapability */
+                       p->jointnoncodeccapability = p->noncodeccapability;
                        ast_debug(2, "No SDP in Invite, third party call control\n");
                }
 
