diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-11.3.0.p008/channels/chan_sip.c asterisk-11.3.0/channels/chan_sip.c
--- asterisk-11.3.0.p008/channels/chan_sip.c	2013-05-20 10:26:13.952937342 +0100
+++ asterisk-11.3.0/channels/chan_sip.c	2013-05-20 11:52:07.913378000 +0100
@@ -11720,7 +11720,7 @@
 		++h;
 	else {
 		h = uri;
-		if (!strncasecmp(h, "sip:", 4)) {
+		if (!strncasecmp(h, "sip:", 4) || !strncasecmp(h, "tel:", 4)) {
 			h += 4;
 		} else if (!strncasecmp(h, "sips:", 5)) {
 			h += 5;
@@ -12412,7 +12412,7 @@
 	ast_copy_string(tmpf, str, sizeof(tmpf));
 	from = get_in_brackets(tmpf);
 	if (!ast_strlen_zero(from)) {
-		if (strncasecmp(from, "sip:", 4)) {
+		if (strncasecmp(from, "sip:", 4) && strncasecmp(from, "tel:", 4)) {
 			ast_log(LOG_WARNING, "Huh?  Not a SIP header (%s)?\n", from);
 			return -1;
 		}
@@ -15822,7 +15822,7 @@
 	ast_copy_string(from, of, sizeof(from));
 	of = get_in_brackets(from);
 	ast_string_field_set(p, from, of);
-	if (!strncasecmp(of, "sip:", 4)) {
+	if (!strncasecmp(of, "sip:", 4) || !strncasecmp(of, "tel:", 4)) {
 		of += 4;
 	} else if (!strncasecmp(of, "sips:", 5)) {
 		of += 5;
@@ -16242,7 +16242,7 @@
 	 * We still need to be able to send to the remote agent through the proxy.
 	 */
 
-	if (parse_uri_legacy_check(contact, "sip:,sips:", &contact, NULL, &hostport,
+	if (parse_uri_legacy_check(contact, "sip:,sips:,tel:", &contact, NULL, &hostport,
 		      &transport)) {
 		ast_log(LOG_WARNING, "Invalid contact uri %s (missing sip: or sips:), attempting to use anyway\n", fullcontact);
 	}
@@ -16378,8 +16378,8 @@
 	ast_string_field_build(pvt, our_contact, "<%s>", curi);
 
 	/* Make sure it's a SIP URL */
-	if (ast_strlen_zero(curi) || parse_uri_legacy_check(curi, "sip:,sips:", &curi, NULL, &hostport, &transport)) {
-		ast_log(LOG_NOTICE, "Not a valid SIP contact (missing sip:/sips:) trying to use anyway\n");
+	if (ast_strlen_zero(curi) || parse_uri_legacy_check(curi, "sip:,sips:,tel:", &curi, NULL, &hostport, &transport)) {
+		ast_log(LOG_NOTICE, "Not a valid SIP contact (missing sip:/sips:/tel:) trying to use anyway\n");
 	}
 
 	/* handle the transport type specified in Contact header. */
@@ -17281,7 +17281,7 @@
 	c = get_in_brackets(tmp);
 	c = remove_uri_parameters(c);
 
-	if (parse_uri_legacy_check(c, "sip:,sips:", &name, &unused_password, &domain, NULL)) {
+	if (parse_uri_legacy_check(c, "sip:,sips:,tel:", &name, &unused_password, &domain, NULL)) {
 		ast_log(LOG_NOTICE, "Invalid to address: '%s' from %s (missing sip:) trying to use anyway...\n", c, ast_sockaddr_stringify_addr(addr));
 		return -1;
 	}
@@ -17714,7 +17714,7 @@
 	if (!end)
 		return 0;
 	*end++ = '\0';
-	if (strncasecmp(start, "sip:", 4))
+	if (strncasecmp(start, "sip:", 4) && strncasecmp(start, "tel:", 4))
 		return 0;
 	cid_num = start + 4;
 	if (global_shrinkcallerid && ast_is_shrinkable_phonenumber(cid_num))
@@ -17789,7 +17789,7 @@
 	}
 
 	exten = get_in_brackets(tmp);
-	if (!strncasecmp(exten, "sip:", 4)) {
+	if (!strncasecmp(exten, "sip:", 4) || !strncasecmp(exten, "tel:", 4)) {
 		exten += 4;
 	} else if (!strncasecmp(exten, "sips:", 5)) {
 		exten += 5;
@@ -18057,7 +18057,7 @@
 
 	uri = ast_strdupa(get_in_brackets(tmp));
 
-	if (parse_uri_legacy_check(uri, "sip:,sips:", &uri, &unused_password, &domain, NULL)) {
+	if (parse_uri_legacy_check(uri, "sip:,sips:,tel:", &uri, &unused_password, &domain, NULL)) {
 		ast_log(LOG_WARNING, "Not a SIP header (%s)?\n", uri);
 		return SIP_GET_DEST_INVALID_URI;
 	}
@@ -18309,7 +18309,7 @@
 	}
 	h_refer_to = ast_strdupa(p_refer_to);
 	refer_to = get_in_brackets(h_refer_to);
-	if (!strncasecmp(refer_to, "sip:", 4)) {
+	if (!strncasecmp(refer_to, "sip:", 4) || !strncasecmp(refer_to, "tel:", 4)) {
 		refer_to += 4;			/* Skip sip: */
 	} else if (!strncasecmp(refer_to, "sips:", 5)) {
 		refer_to += 5;
@@ -18335,7 +18335,7 @@
 
 		referred_by_uri = get_in_brackets(h_referred_by);
 
-		if (!strncasecmp(referred_by_uri, "sip:", 4)) {
+		if (!strncasecmp(referred_by_uri, "sip:", 4) || !strncasecmp(referred_by_uri, "tel:", 4)) {
 			referred_by_uri += 4;		/* Skip sip: */
 		} else if (!strncasecmp(referred_by_uri, "sips:", 5)) {
 			referred_by_uri += 5;		/* Skip sips: */
@@ -18497,7 +18497,7 @@
 	ast_copy_string(tmp, sip_get_header(req, "Also"), sizeof(tmp));
 	c = get_in_brackets(tmp);
 
-	if (parse_uri_legacy_check(c, "sip:,sips:", &c, NULL, &a, NULL)) {
+	if (parse_uri_legacy_check(c, "sip:,sips:,tel:", &c, NULL, &a, NULL)) {
 		ast_log(LOG_WARNING, "Huh?  Not a SIP header in Also: transfer (%s)?\n", c);
 		return -1;
 	}
@@ -18943,7 +18943,7 @@
 
 	if (ast_strlen_zero(p->exten)) {
 		char *t = uri2;
-		if (!strncasecmp(t, "sip:", 4))
+		if (!strncasecmp(t, "sip:", 4) || !strncasecmp(t, "tel:", 4))
 			t+= 4;
 		else if (!strncasecmp(t, "sips:", 5))
 			t += 5;
@@ -18961,7 +18961,7 @@
 	/* save the URI part of the From header */
 	ast_string_field_set(p, from, of);
 
-	if (parse_uri_legacy_check(of, "sip:,sips:", &name, &unused_password, &domain, NULL)) {
+	if (parse_uri_legacy_check(of, "sip:,sips:,tel:", &name, &unused_password, &domain, NULL)) {
 		ast_log(LOG_NOTICE, "From address missing 'sip:', using it anyway\n");
 	}
 
@@ -22907,7 +22907,7 @@
 
 	if (set_call_forward && ast_test_flag(&p->flags[0], SIP_PROMISCREDIR)) {
 		char *host = NULL;
-		if (!strncasecmp(contact_number, "sip:", 4))
+		if (!strncasecmp(contact_number, "sip:", 4) || !strncasecmp(contact_number, "tel:", 4))
 			contact_number += 4;
 		else if (!strncasecmp(contact_number, "sips:", 5))
 			contact_number += 5;
@@ -22937,7 +22937,7 @@
 		if (separator)
 			*separator = '\0';
 
-		if (!strncasecmp(contact_number, "sip:", 4))
+		if (!strncasecmp(contact_number, "sip:", 4) || !strncasecmp(contact_number, "tel:", 4))
 			contact_number += 4;
 		else if (!strncasecmp(contact_number, "sips:", 5))
 			contact_number += 5;
diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-11.3.0.p008/channels/sip/reqresp_parser.c asterisk-11.3.0/channels/sip/reqresp_parser.c
--- asterisk-11.3.0.p008/channels/sip/reqresp_parser.c	2013-05-20 10:26:14.444487857 +0100
+++ asterisk-11.3.0/channels/sip/reqresp_parser.c	2013-05-20 14:10:46.220836000 +0100
@@ -100,10 +100,16 @@
 			dom = c;
 			userinfo = uri;
 			uri = c; /* userinfo can contain ? and ; chars so step forward before looking for params and headers */
+		}
+		else if ( (c = strstr(uri, ";phone-context=")) ) { /* TEL URI INVITE */
+		   *c = '\0';
+		   dom = c + 15;
+		   userinfo = uri;
+		   uri = dom;
 		} else {
-			/* domain-only URI, according to the SIP RFC. */
-			dom = uri;
-			userinfo = "";
+		   /* domain-only URI, according to the SIP RFC. */
+		   dom = uri;
+		   userinfo = "";
 		}
 
 		*hostport = dom;
@@ -1306,7 +1312,7 @@
 		params.transport = params.user = params.method = params.ttl = params.maddr = NULL;
 		params.lr = 0;
 		ast_copy_string(uri,testdataptr->uri,sizeof(uri));
-		if (parse_name_andor_addr(uri, "sip:,sips:",
+		if (parse_name_andor_addr(uri, "sip:,sips:,tel:",
 					  &name,
 					  &user,
 					  &pass,
