diff -Naur orig_asterisk-i-1.8.3.2/codecs/amr/typedef.h asterisk-i-1.8.3.2/codecs/amr/typedef.h
--- orig_asterisk-i-1.8.3.2/codecs/amr/typedef.h	2012-07-05 16:13:48.000000000 +0100
+++ asterisk-i-1.8.3.2/codecs/amr/typedef.h	2012-07-03 11:00:18.000000000 +0100
@@ -15,7 +15,7 @@
 typedef char Word8;
 typedef unsigned char UWord8;
 typedef short Word16;
-typedef long Word32;
+typedef int Word32;
 typedef float Float32;
 typedef double Float64;
 
