diff --exclude='*.xml' --exclude='*.o.d' --exclude=.project --exclude=.cproject -Naur asterisk-base-11.3.0/apps/wms_inoapi.c asterisk-11.3.0/apps/wms_inoapi.c
--- asterisk-base-11.3.0/apps/wms_inoapi.c	2022-06-29 14:41:17.308889617 +0100
+++ asterisk-11.3.0/apps/wms_inoapi.c	2022-06-29 15:24:16.739910811 +0100
@@ -625,13 +625,6 @@
             ast_log(LOG_WARNING, "Thread creation ERROR |%s| for conference %d on ch %03d - %s\n", strerror(errno), trs_cmd.al_int_val2, trs_cmd.ch_index, ast_channel_name(chan));
             goto error;
          }
-         else
-         {
-            if (option_verbose >= 4)
-               ast_verbose(VERBOSE_PREFIX_4 "User added to conference %d|%s on ch %03d - %s\n",
-                  trs_cmd.al_int_val2, trs_cmd.al_char_val1, trs_cmd.ch_index, ast_channel_name(chan));
-            asteriskSendReply(&trs_reply, chan_vars);
-         }
       }
       else
       {
@@ -912,12 +905,17 @@
       if (option_verbose >= 3)
          ast_verbose(VERBOSE_PREFIX_3 "Conference_thread created |%s| on channel %s\n", execCmd, ast_channel_name(chan));
 
+      trs_reply.body.g_ret_value = 1;
+      asteriskSendReply(&trs_reply, chan_vars);
+ 
       ast_channel_ref(chan); // when calling a new app is necessary to increase the channel ref
       pbx_exec(chan, app, execCmd);
       ast_channel_unref(chan);
 
-      if (chan_vars->conference.exit)
+      if (chan_vars->conference.exit){
+         trs_reply.body.g_ret_value = 2;
          asteriskSendReply(&trs_reply, chan_vars);
+      }
       else
       {
          trs_reply.event = UNSOLICITED_EVENT;
