diff -Naur asterisk-base-1.8.3.2/formats/format_pcm.c asterisk-1.8.3.2/formats/format_pcm.c
--- asterisk-base-1.8.3.2/formats/format_pcm.c	2012-08-28 13:59:13.000000000 +0100
+++ asterisk-1.8.3.2/formats/format_pcm.c	2012-08-28 14:01:07.000000000 +0100
@@ -415,7 +415,7 @@
 
 static const struct ast_format alaw_f = {
 	.name = "alaw",
-	.exts = "alaw|al|alw",
+	.exts = "alaw|al|alw|pcm|vox",
 	.format = AST_FORMAT_ALAW,
 	.write = pcm_write,
 	.seek = pcm_seek,
@@ -432,7 +432,7 @@
 
 static const struct ast_format pcm_f = {
 	.name = "pcm",
-	.exts = "pcm|ulaw|ul|mu|ulw",
+	.exts = "ulaw|ul|mu|ulw",
 	.format = AST_FORMAT_ULAW,
 	.write = pcm_write,
 	.seek = pcm_seek,
