diff -Naur orig_asterisk-i-1.8.3.2/channels/sip/reqresp_parser.c asterisk-i-1.8.3.2/channels/sip/reqresp_parser.c
--- orig_asterisk-i-1.8.3.2/channels/sip/reqresp_parser.c	2010-07-28 21:49:26.000000000 +0100
+++ asterisk-i-1.8.3.2/channels/sip/reqresp_parser.c	2012-07-03 10:53:02.000000000 +0100
@@ -75,6 +75,13 @@
 			dom = c;
 			userinfo = uri;
 			uri = c; /* userinfo can contain ? and ; chars so step forward before looking for params and headers */
+		
+		} else if ( (c = strstr(uri, ";phone-context=")) ) { /* TEL URI INVITE */
+			*c = '\0';
+			dom = c + 15;
+			userinfo = uri;
+			uri = dom;
+		
 		} else {
 			/* domain-only URI, according to the SIP RFC. */
 			dom = uri;
@@ -448,7 +455,7 @@
 		params.lr = 0;
 
 		ast_copy_string(uri,testdataptr->uri,sizeof(uri));
-		if (parse_uri_full(uri, "sip:,sips:", testdataptr->userptr,
+		if (parse_uri_full(uri, "sip:,sips:,tel:", testdataptr->userptr,
 				   testdataptr->passptr, testdataptr->domainptr,
 				   testdataptr->paramsptr,
 				   testdataptr->headersptr,

